lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp

```diff
@@ -1592,40 +1592,34 @@ LogicalResult FIRRTLLowering::visitDecl(MemOp op) {
       // Add delays for non-zero read latency
       SmallVector<ReadPipeElement> readPipe;
       readPipe.push_back({portWires["en"], portWires["addr"]});
-      Value enReg, addrReg;
-      for (size_t j = 0; j < readLatency; ++j) {
-        if (j == 0) {
-          enReg = arrayReg("_en_pipe", i1Type, readLatency);
-          addrReg = arrayReg("_addr_pipe", lowerType(addrType), readLatency);
+      if (readLatency > 0) {
+        Value enReg = arrayReg("_en_pipe", i1Type, readLatency);
+        Value addrReg =
+            arrayReg("_addr_pipe", lowerType(addrType), readLatency);
+        for (size_t j = 0; j < readLatency; ++j) {
+          auto jIdx =
+              builder->create<rtl::ConstantOp>(APInt(log2(readLatency + 1), j));
+          auto enJ = builder->create<sv::ArrayIndexInOutOp>(enReg, jIdx);
+          auto addrJ = builder->create<sv::ArrayIndexInOutOp>(addrReg, jIdx);
+          readPipe.push_back({enJ, addrJ});
+        }
+        for (size_t j = 1; j <= readLatency; ++j) {
+          auto lastEn = readInOut(readPipe[j - 1].en);
+          auto lastAddr = readInOut(readPipe[j - 1].addr);
+          addToAlwaysFFBlock(EventControl::AtPosEdge, clk, [&]() {
+            builder->create<sv::PAssignOp>(readPipe[j].en, lastEn);
+            builder->create<sv::PAssignOp>(readPipe[j].addr, lastAddr);
+          });
         }
-        auto jIdx =
-            builder->create<rtl::ConstantOp>(APInt(log2(readLatency + 1), j));
-        auto enJ = builder->create<sv::ArrayIndexInOutOp>(enReg, jIdx);
-        auto addrJ = builder->create<sv::ArrayIndexInOutOp>(addrReg, jIdx);
-        readPipe.push_back({enJ, addrJ});
-      }
-      if (readLatency != 0) {
-        addToAlwaysFFBlock(EventControl::AtPosEdge, clk, [&]() {
-          auto lastEn = readPipe[0].en;
-          auto lastAddr = readPipe[0].addr;
-          for (size_t j = 1; j < readLatency + 1; ++j) {
-            auto thisEn = readPipe[j].en;
-            auto thisAddr = readPipe[j].addr;
-
-            auto readLastEn = builder->create<sv::ReadInOutOp>(lastEn);
-            builder->create<sv::PAssignOp>(thisEn, readLastEn);
-            builder->create<sv::IfOp>(readLastEn, [&]() {
-              builder->create<sv::PAssignOp>(
-                  thisAddr, builder->create<sv::ReadInOutOp>(lastAddr));
-            });
-            lastEn = thisEn;
-            lastAddr = thisAddr;
-          }
-        });
       }
-      Value addr = builder->create<sv::ReadInOutOp>(readPipe.back().addr);
-      Value value = builder->create<sv::ArrayIndexInOutOp>(reg, addr);
-      value = builder->create<sv::ReadInOutOp>(value);
+      Value addr = readInOut(readPipe.back().addr);
+      Value en = readInOut(readPipe.back().en);
+      Value value =
+          readInOut(builder->create<sv::ArrayIndexInOutOp>(reg, addr));
+      // FIRRTL 5.11.1 says undefined behavior on low enable.
+      auto undefinedVal =
+          builder->create<sv::TextualValueOp>(value.getType(), "'x");
```

Comment:
Unrelated to this patch, but we should have an `sv::UnspecifiedValue` op that is analogous to Constant that produces X's, same thing for `'z`.

Reply:
I agree that this is needed. It may warrant additional discussion due to the weirdness of Verilog `x` and if there is a better type hierarchy present. Possible interpretations of `x` all include: "free for the compiler to optimize", "undefined hardware behavior", "poison", "uninitialized", "a unique source of randomness", and "the same source of randomness".

Reply:
I was going to add a constant type for 'x' to sv.

Reply:
Thanks!  I [filed an issue](https://github.com/llvm/circt/issues/725) to track this. @seldridge I think that `x` needs to follow the behavior and expectations of widely used tools + xprop, we don't have full freedom on how to define it.  We can also add an undef/poison node with more freedom if there is a reason to do so.